### PR TITLE
Fix null handling in SortedPositionLinks

### PR DESCRIPTION
**Pull request by sopel39**
Fix for: https://github.com/prestodb/presto/issues/8025

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java [hunk 1]
```diff
@@ -52,22 +55,35 @@
...
         }
 
         @Override
         public int link(int from, int to)
         {
+            if (isNull(from)) {
+                return to;
+            }
+
+            if (isNull(to)) {
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java]: what if both are nulls or all of the values are null? Do you have tests for that?

File: presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java [hunk 2]
```diff
@@ -52,22 +55,35 @@
...
+            this.comparator = comparator;
+            this.pagesHashStrategy = pagesHashStrategy;
+            this.addresses = addresses;
             positionLinks = new Int2ObjectOpenHashMap<>();
         }
 
         @Override
         public int link(int from, int to)
         {
+            if (isNull(from)) {
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java]: what if there is only one null value? Do you have tests for that?

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java]: added test: `SELECT b FROM nation n, (VALUES (0, NULL)) t(a, b) WHERE n.regionkey - 100 < t.b AND n.nationkey = t.a`

File: presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java]: added test: `SELECT b FROM nation n, (VALUES (0, NULL), (0, NULL), (0, CAST(-1 AS BIGINT)), (0, NULL)) t(a, b) WHERE n.regionkey - 100 < t.b AND n.nationkey = t.a`

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 3]
```diff
@@ -2121,6 +2121,13 @@ public void testJoinWithLessThanInJoinClause()
     {
         assertQuery("SELECT n.nationkey, r.regionkey FROM region r JOIN nation n ON n.regionkey = r.regionkey AND n.name < r.name");
         assertQuery("SELECT l.suppkey, n.nationkey, l.partkey, n.regionkey FROM nation n JOIN lineitem l ON l.suppkey = n.nationkey AND l.partkey < n.regionkey");
+        assertQuery("SELECT b FROM nation n, (VALUES (0, CAST(-1 AS BIGINT)), (0, NULL), (0, CAST(0 AS BIGINT))) t(a, b) WHERE n.regionkey - 100 < t.b AND n.nationkey = t.a",
```
Comment (dain) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Add comments like `// test with single NULL in build`

File: presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java]: added TODO

**Comment by sopel39** [comment]
@dain this is my commit explanation:
```
    Previously null rows were added to SortedPositionLinks chain.
    Block value for null row is undefined, therefore null rows
    hadn't had any specified position in sorted chain. This caused
    sorted position links enumeration to prematurely terminate
    (since join filter function returns false for null row) without
    visiting all rows for which filter function returns true.
    
    This fix causes null rows not to be added to SortedPositionLinks
    since nulls rows will always be filtered out by join filter function.
    When two positions are linked using SortedPositionLinks.Builder#link
    a check is made if for any of the two rows a sort channel value is null.
    If it is, then the other (possibly not null) row is returned as
    a chain head. Therefore null row can only be part of a chain if there
    are no other non-null rows.


```

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 3 above)
Comment (sopel39) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: added comments

**Comment by dain** [comment]
Merged, thanks!

**Pull request closed without merging**